src/madsci_common/madsci/common/mongodb_migration_tool.py

```diff
@@ -0,0 +1,610 @@
+            current_dir / "data_manager" / "schema.json",
+            current_dir / "schema" / "data_manager.json",
+        ]
+    else:
+        # Generic paths for custom database names
+        possible_paths = [
+            current_dir / "madsci" / database_name / "schema.json",
+            current_dir / database_name / "schema.json",
+            current_dir / "schema" / f"{database_name}.json",
+        ]
+
+    # Find the first existing schema file
+    for path in possible_paths:
+        if path.exists():
+            return path
+
+    # If no schema file found, provide helpful error
+    paths_str = "\n".join(f"  - {path}" for path in possible_paths)
+    raise FileNotFoundError(
+        f"No schema file found for database '{database_name}'. "
+        f"Searched in:\n{paths_str}\n\n"
+        f"Please either:\n"
+        f"1. Create a schema.json file in one of the above locations, or\n"
+        f"2. Specify the path explicitly with --schema-file"
+    )
+
+
+def setup_migration_components(
+    db_url: str, database: str, schema_file_path: Path, logger: EventClient
+) -> tuple[MongoDBVersionChecker, MongoDBMigrator]:
+    """Setup version checker and migrator components."""
+    version_checker = MongoDBVersionChecker(
+        db_url=db_url,
+        database_name=database,
+        schema_file_path=str(schema_file_path),
+        logger=logger,
+    )
+
+    migrator = MongoDBMigrator(
+        db_url=db_url,
+        database_name=database,
+        schema_file_path=str(schema_file_path),
+        logger=logger,
+    )
+
+    return version_checker, migrator
+
+
+def handle_migration_commands(
+    args: Any,
+    version_checker: MongoDBVersionChecker,
+    migrator: MongoDBMigrator,
+    logger: EventClient,
+) -> None:
+    """Handle different migration command options."""
+    if args.check_version:
+        # Just check version compatibility
+        needs_migration, madsci_version, db_version = (
+            version_checker.is_migration_needed()
+        )
+
+        logger.log_warning(f"MADSci version: {madsci_version}")
+        logger.log_warning(f"Database version: {db_version or 'None'}")
+        logger.log_warning(f"Migration needed: {needs_migration}")
+
+        if needs_migration:
+            sys.exit(1)  # Exit with error code if migration needed
+
+    elif args.restore_from:
+        # Restore from backup
+        backup_path = Path(args.restore_from)
+        migrator.restore_from_backup(backup_path)
+        logger.info("Restore completed successfully")
+
+    elif args.backup_only:
+        # Just create backup
+        backup_path = migrator.create_backup()
+        logger.info(f"Backup created: {backup_path}")
+
+    else:
+        # Run full migration
+        migrator.run_migration(args.target_version)
+        logger.info("Migration completed successfully")
+
+
+def main() -> None:
+    """Command line interface for the MongoDB migration tool."""
+    parser = argparse.ArgumentParser(
```

Comment:
Broken record at this point, but we should replace the argparse with PydanticSettings model and take advantage of https://docs.pydantic.dev/latest/concepts/pydantic_settings/#command-line-support